### PR TITLE
test: add unit tests for BaseStrategy using bulloak

### PR DESCRIPTION
**Pull request by 0xAustrian**

### CODE REVIEW

File: test/unit/strategies/BaseStrategyUnit.tree [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+├── It should add amount to poolAmount
+└── It should call _afterIncreasePoolAmount
+
+BaseStrategy::withdraw
+├── When parameters are valid
+│   ├── It should call onlyPoolManager
+│   ├── It should call _beforeWithdraw
+│   ├── It should call balanceOf at token
+│   ├── It should call transfer at token
+│   ├── It should call _afterWithdraw
```
Comment (0xOneTony) [inline test/unit/strategies/BaseStrategyUnit.tree:21]: lets also add maybe, `It should not change the pool amount`

File: test/unit/strategies/BaseStrategyUnit.t.sol [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+        baseStrategy.mock_call__checkOnlyAllo();
+
+        // It should call _afterIncreasePoolAmount
+        baseStrategy.expectCall__afterIncreasePoolAmount(_amount);
+
+        baseStrategy.increasePoolAmount(_amount);
+    }
+
+    function test_WithdrawWhenParametersAreValid(address _token, uint256 _amount, address _recipient) external {
+        // TODO: Fix the bug and fix this test
```
Comment (0xOneTony) [inline test/unit/strategies/BaseStrategyUnit.t.sol]: What is the issue here?

Comment (0xAustrian) [reply test/unit/strategies/BaseStrategyUnit.t.sol]: I'm referring to [this](https://github.com/allo-protocol/allo-v2.1/pull/17) bug

File: test/unit/strategies/BaseStrategyUnit.tree (hunk 1 above)
Comment (0xAustrian) [reply test/unit/strategies/BaseStrategyUnit.tree:21]: Sure, will add it after #17 is merged

**Review by 0xOneTony — APPROVED** [review]

**Review by thelostone-mc — APPROVED** [review]

File: test/unit/strategies/BaseStrategyUnit.tree [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+│   ├── It should call onlyPoolManager
+│   ├── It should call _beforeWithdraw
+│   ├── It should call getPool at allo
+│   ├── It should call transfer at token
+│   ├── It should call _afterWithdraw
+│   ├── It should not change the pool amount
+│   └── It should emit event
+└── When token is pool token
+    ├── It should call balanceOf at token
+    └── When amount is greater than poolAmount
```
Comment (fnanni-0) [inline test/unit/strategies/BaseStrategyUnit.tree:26]:
```suggestion
    └── When amount is greater than the balance minus poolAmount
```

File: test/unit/strategies/BaseStrategyUnit.t.sol [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+        vm.assume(_amount > 0);
+        vm.assume(_contractBalance > _amount);
+
+        baseStrategy.set__poolAmount(_contractBalance);
+
+        // It should revert
+        vm.expectRevert(IBaseStrategy.BaseStrategy_WITHDRAW_MORE_THAN_POOL_AMOUNT.selector);
+
+        baseStrategy.withdraw(_token, _amount, _recipient);
+    }
```
Comment (fnanni-0) [inline test/unit/strategies/BaseStrategyUnit.t.sol:217]: There should be three important variables in this test: poolAmount, balance and _amount. What do you think of something like this?
```suggestion
    function test_WithdrawRevertWhen_AmountIsGreaterThanPoolAmount(
        address _token,
        uint256 _amount,
        address _recipient,
        uint256 _contractBalance,
        uint256 _poolAmount
    ) external whenTokenIsPoolToken(_token, _amount, _recipient, _contractBalance) {
        _poolAmount = bound(_poolAmount, 1, _contractBalance);
        _amount = bound(_amount, _contractBalance - _poolAmount + 1, _contractBalance);

        baseStrategy.set__poolAmount(_poolAmount);

        // It should revert
        vm.expectRevert(IBaseStrategy.BaseStrategy_WITHDRAW_MORE_THAN_POOL_AMOUNT.selector);

        baseStrategy.withdraw(_token, _amount, _recipient);
    }
```

Comment (ilpepepig) [inline test/unit/strategies/BaseStrategyUnit.t.sol]: There should be three important variables in this test: poolAmount, balance and _amount. What do you think of something like this?
```suggestion
    function test_WithdrawRevertWhen_AmountIsGreaterThanPoolAmount(
        address _token,
        uint256 _amount,
        address _recipient,
        uint256 _contractBalance,
        uint256 _poolAmount
    ) external whenTokenIsPoolToken(_token, _amount, _recipient, _contractBalance) {
        _poolAmount = bound(_poolAmount, 1, _contractBalance);
        _amount = bound(_amount, _contractBalance - _poolAmount + 1, _contractBalance);

        baseStrategy.set__poolAmount(_poolAmount);

        // It should revert
        vm.expectRevert(IBaseStrategy.BaseStrategy_WITHDRAW_MORE_THAN_POOL_AMOUNT.selector);

        baseStrategy.withdraw(_token, _amount, _recipient);
    }
```

File: test/unit/strategies/BaseStrategyUnit.tree (hunk 3 above)
Comment (ilpepepig) [inline test/unit/strategies/BaseStrategyUnit.tree]:
```suggestion
    └── When amount is greater than the balance minus poolAmount
```

**Review by ilpepepig — APPROVED** [review]

**Pull request merged**